keanu-project/src/main/java/io/improbable/keanu/util/dot/WriteDot.java

```diff
@@ -0,0 +1,279 @@
+import java.io.IOException;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.stream.Stream;
+
+/**
+ * Utility class for outputting a network to a DOT file that can then be used by a range of graph visualisers.
+ * Read more about DOT format here: https://en.wikipedia.org/wiki/DOT_(graph_description_language)
+ *
+ * Usage:
+ * To output network to a DOT file: WriteDot.outputDot(fileName, network)
+ * To output vertex and it's connections up to degree n: WriteDot.outputDot(fileName, vertex, n)
+ */
+public class WriteDot {
+
+    private static BufferedWriter writer;
+    private static HashMap<Integer, String> idsToLabelStrings = new HashMap<>();
+    private static HashMap<String, String> verticeIdsToConnectionInfo = new HashMap<>();
+
+
+    /**
+     * Annotation used for vertex classes to specify how they should be exported to a DOT file.
+     */
+    @Retention(RetentionPolicy.RUNTIME)
+    public @interface DotAnnotation {
+        String displayName() default "";
+        boolean displayHyperparameterInfo() default false;
+    }
+
+
+    /**
+     * Outputs the given graph to a DOT file which can be used by various graph visualisers to generate a visual representation of the graph.
+     * Read more about DOT format here: https://en.wikipedia.org/wiki/DOT_(graph_description_language)
+     *
+     * @param fileName name of the output file.
+     * @param net network to be written out to DOT format.
+     */
+    public static void outputDot(String fileName, BayesianNetwork net) {
+
+        // Get any vertex in the network and set the degree to infinity to print out the entire network.
+        if (net.getAllVertices().size() == 0) return;
+        Vertex someVertex = net.getAllVertices().get(0);
+        outputDot(fileName, someVertex, Integer.MAX_VALUE);
+    }
+
+
+    /**
+     * Outputs the given graph to a DOT file which can be used by various graph visualisers to generate a visual representation of the graph.
+     * Read more about DOT format here: https://en.wikipedia.org/wiki/DOT_(graph_description_language)
+     *
+     * @param fileName name of the output file
+     * @param vertex vertex around which the graph will be visualised
+     * @param degree degree of connections to be visualised; for instance, if the degree is 1,
+     *               only connections between the vertex amd it's parents and children will be written out to the DOT file.
+     */
+    public static void outputDot(String fileName, Vertex vertex, int degree) {
+        try {
+            // Reset variables
+            idsToLabelStrings = new HashMap<>();
+            verticeIdsToConnectionInfo = new HashMap<>();
+
+            File outputFile = getOutputFile(fileName);
+
+            writer = new BufferedWriter(new FileWriter(outputFile));
+            writer.write("digraph BayesianNetwork {\n");
```

Comment:
I think the code here would be a bit more self describing if we used either:
a) A separate "DotOutputStream" kind of class (that would wrap our real output stream object and, for instance, stick this header on automatically).
b) Some "final" variables describing what they are (ie like public static final String DOT_HEADER = "digr...

Reply:
Makes sense, went for b)